### ISSUE TITLE
Examine HTML5 Audio support in Google Play

### ISSUE THREAD

**Issue by fenryxo** [tiliado/nuvolaplayer#91]
HTML 5 Audio used to work in Nuvola Player, but it is no longer true. Examine why.

Chromium can play without Flash plugin.

**Comment by mcg** [comment]
I know $15 isn't much, but I can't use Nuvola anymore as I've removed Flash from all my systems.

**Comment by fenryxo** [comment]
You could try Rdio. It supports HTML5 audio nicely.

**Comment by mcg** [comment]
Kind of invested in Google Play. HTML5 playback works in Chrome, but I am guessing it's, some DRM module Chrome is shipping.

**Comment by fenryxo** [comment]
Related ticket in Radiant Player: kbhomes/radiant-player-mac#335

**Comment by fenryxo** [comment]
HTML5 Audio seems to work in Nuvola Player 3.1 in Debian Jessie for songs in the library, not sure about All Access Music subscription. Can anybody interested investigate? First, disable the Flash plugin in App menu/Gear menu → Format support → Web plugins → Toggle "Load web plugins (Flash)" off, then restart Nuvola Player.

**Comment by jnv** [comment]
I have Nuvola Player 3.0.1 and All Access; unfortunately I cannot build version 3.1 now. With Flash Player disabled, Google Play insists on installing Flash Player and refuses to play anything (even my uploads).

However apparently it can be done with HTML5 audio. Google Play Music Desktop Player recently switched to Electron got rid of Flash Player in the process: https://github.com/MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-/issues/60

**Comment by fenryxo** [comment]
Thanks @jnv. However, I'm particularly interested in a feedback about NP 3.1 as it contains, unlike NP 3.0.1, some tweaks for better support of HTML5 Audio.

**Comment by KruyKaze** [comment]
I'm using 3.1 i deactivated flash and restarted. Still complains about not
having flash for music in the library and can't toggle html5 in google
music settings.
Arch 64
Thanks

**Comment by fenryxo** [comment]
**HTML5 Audio** in Google Play Music works in **Nuvola Player 3.1** in Ubuntu 14.04, 15.10 and 16.04, Debian Jessie and Fedora 23 with my account **without All Access subscription**. I'm going to check also Ubuntu 15.04, Fedora 22 and Arch Linux later and post here more details. Stay tuned :-)

**Comment by bartlibert** [comment]
Just confirming that HTML5 audio is not working with 3.1 in Debian Sid with All Access

**Comment by Celti** [comment]
HTML5 audio also fails with git HEAD on Arch Linux with All Access. In fact, the HTML5 toggle is greyed out in every browser I have access to. HTML5 audio works fine in Chromium, even with all plugins disabled (Pepper Flash, Native Client, Widevine Content Decryptor), fails in Firefox.

**Comment by fenryxo** [comment]
:heart: Another [Nuvola Patron](https://tiliado.eu/nuvolaplayer/funding/) is interested in this feature.

**Comment by fenryxo** [comment]
Update on current progress:

* HTML5 Audio works in Nuvola Player 3.1 in all supported distributions but only with accounts without All Access subscription. Media Source Extension (MSE) and Chrome's user agent are probably necessary for complete support. Not sure about Encrypted Media Extension (EME) though.
* Let's focus only on the latest distributions: Ubuntu 16.04 & Arch Linux at first, Debian Stretch/Sid & Fedora 24 later.
* My Launchpad PPA [Experimental WebKitGTK builds for Nuvola Player](https://launchpad.net/~nuvola-player-builders/+archive/ubuntu/webkit-experimental) contains WebKitGTK+ 2.12.2 with MSE enabled (Ubuntu 16.04 only). It should be possible to create an AUR PKGBUILD with `-DENABLE_MEDIA_SOURCE=ON` for Arch Linux (@Celti ?).
* Nuvola Player 3.1 supports [a custom user agent string](https://github.com/tiliado/nuvolaplayer/commit/3afde6fd6438ec28ef0457524f43a1f2dbf5a8d6).
* However, if [the user agent string is set to Chrome](https://github.com/tiliado/nuvola-app-google-play-music/commit/1d918c51cbea35ea9b3fe1688c2fdbdce877676f), Google Play Music fails to play with  the `CustomError: Error in protected function: null is not an object (evaluating 'this.zxa.aZa')` console error, which needs to be investigated further.

To be continued in July...

**Comment by Celti** [comment]
I'm quite sure EME is not necessary — Google Play Music with All Access works fine in open-source Chromium without the Widevine EME (or any EME, in fact) installed.

I'll get rolling on a custom WebKit package in an hour or so.

**Comment by Celti** [comment]
webkit2gtk-mse-2.12.3-1 is in the repository. With it, latest Nuvola Player 3.1 and Google Play still demand Flash, with or without a Chrome user-agent.

Interesting to note that Google seems to have completely removed the HTML5 toggle from the Play Music settings menu.

**Comment by fenryxo** [comment]
> Geez, this monster is slow to build! I think I'll only build it for x86_64 for the moment, unless and until it turns out it's necessary for HTML5 (and then I'll submit a request upstream to get that flag in the Arch package).
> ETA 2: Aaaaugh, churned its way up to 79% and then the OOM-killer reaped g++! Time to run the build locally.

It might help to use [`-g1` instead of `-g` in `CFLAGS`](https://github.com/tiliado/webkitgtk-pkg-deb/commit/5911350db3f013aefde3cd285a2ab4627ce9e7e8).

> webkit2gtk-mse-2.12.3-1 is in the repository. With it, latest Nuvola Player 3.1 and Google Play still demand Flash, with or without a Chrome user-agent.

Could you post [debugging output](https://github.com/tiliado/nuvolaplayer/wiki/Bug-Reporting-Guidelines#debugging-output) with [the html5-audio branch](https://github.com/tiliado/nuvola-app-google-play-music/tree/html5-audio)?

> Interesting to note that Google seems to have completely removed the HTML5 toggle from the Play Music settings menu.

Yes, HTML5 Audio is currently used by default instead of Flash if all necessary JavaScript API and codecs are detected.

**Comment by Celti** [comment]
Debug output: https://gist.github.com/Celti/4201c2a22e52a8b71a29c35bf83bd011
I tried playing both a regular station and files I'd uploaded to Google Play.

I wonder if the `"Audio.canPlayType ('audio/mpeg') => maybe"` is making it choke, even if it determines "probably" for actual MP3 support farther down?

**Comment by fenryxo** [comment]
> I wonder if the `"Audio.canPlayType ('audio/mpeg') => maybe"` is making it choke, even if it determines "probably" for actual MP3 support farther down?

Indeed, Audio.canPlayType seems to return "maybe" if a codec is not specified.

```
MIME type	<audio> tag	Audio() object	Media Source
audio/ogg	maybe	maybe	null
audio/mpeg	maybe	maybe	null
audio/mpeg; codecs="flump3dec"	probably	probably	null
audio/mpeg; codecs="mp3"	probably	probably	null
audio/x-wav	maybe	maybe	null
audio/basic	no	no	null
audio/x-aiff	maybe	maybe	null
video/webm	maybe	maybe	null
video/webm; codecs="vorbis,vp8"	probably	probably	null
```

I've just added a small hack to the html5-audio branch but I haven't checked it with mse-enabled webkit yet.

**Comment by Celti** [comment]
Sorry for the long delay in getting back to this. [Here](https://gist.github.com/ce4845bc3f25e8550b8856bc3da084e3) is a debug log with the latest nuvolaplayer 3.1 build with MSE and the html5-audio branch.

Behaviour: takes forever to load the initial page; spins endlessly when I try and play a song; rapidly tries every song in the playlist then fails with "Can't reach Google Play, please reload the page" when I hit skip.

**Comment by fenryxo** [comment]
Hello @Celti. So we have now tricked GPM to think MSE is supported. That's the first step. Unfortunately, the MSE support in WebKitGTK is incomplete. There is a fork with a full MSE support somewhere on GitHub (I can't find the link right now though) and not all patches from it has landed to the upstream WebKitGTK repo yet. I could try to contact the author as soon as I find the link.

**Comment by fenryxo** [comment]
https://github.com/eocanha/webkit/commits/mse-backport

**Comment by fenryxo** [comment]
There is a lot of activity on [the upstream ticket [GStreamer][MSE] Complete backend rework](https://bugs.webkit.org/show_bug.cgi?id=157314) 🎉

**Comment by fenryxo** [comment]
Related: Create Nuvola CDK with MSE turned on #288

**Comment by fenryxo** [comment]
Upstream issue: [ [GTK] [MSE] Turn MSE on by default](https://bugs.webkit.org/show_bug.cgi?id=167107)

**Comment by fenryxo** [comment]
Hello Google Play Music fans. I've managed to build the [Nuvola Core Developer Kit](https://github.com/tiliado/nuvolaplayer/wiki/Nuvola-Core-Developer-Kit) flatpak with Media Source Extension. That's the technology behind a Flash-free playback in Chrome. However, since I don't have Google Play Music subscription, I cannot tell whether the current status of MSE support in WebKitGTK is sufficient. Is there anybody who would like to help me with testing? Basic experience with Linux command line is required though.

**Comment by excieve** [comment]
@fenryxo I'd be happy to test it. I'll be trying a flatpak build for another issue anyway.

**Comment by fenryxo** [comment]
> I'd be happy to test it. I'll be trying a flatpak build for another issue anyway.

That's a different flatpak build - eu.tiliado.Nuvola flatpak is Nuvola Player 3.1 with the stable WebKitGTK 2.14 (without MSE). MSE needs to be tested with eu.tiliado.NuvolaCoreDeveloperKit - a development environment with a development release of WebKitGTK 2.15.4 (with MSE).

Anyway, I'll provide you with instructions within a week as I still need to do a few tests on my own.

**Comment by fenryxo** [comment]
Hello @excieve. MSE support in Nuvola ADK/CDK is not ready yet. I've figured out that GStreamer libs must be upgraded (#288).

**Comment by fenryxo** [comment]
There is now an experimental build of Google Play Music with MSE enabled. Is anybody willing to test it?

**Comment by excieve** [comment]
@fenryxo I'd be glad to help. How do I install it?

**Comment by fenryxo** [comment]
1. Run `flatpak update && flatpak update --user'.
2. Follow installation instructions of [Nuvola 4](https://nuvola.tiliado.eu/nuvola/).
3. Find out the name of Nuvola remote: `flatpak remote-list` → e.g. `eu.tiliado.Nuvola-origin`
4. Install the master branch of Nuvola Runtime: `flatpak install eu.tiliado.Nuvola-origin eu.tiliado.Nuvola//master`
5. Install the master branch of Google Play Music: `flatpak install eu.tiliado.Nuvola-origin eu.tiliado.NuvolaAppGooglePlayMusic//master`
6. Mark the master branch as active `su -c 'flatpak make-current eu.tiliado.Nuvola master'; su -c 'flatpak make-current eu.tiliado.NuvolaAppGooglePlayMusic master'`
7. Launch Google Play Music: `flatpak run eu.tiliado.NuvolaAppGooglePlayMusic//master -D`

**Comment by excieve** [comment]
When starting fresh it crashes to a blank view when the playback is about to start. The following is in the debug log:
```
Worker: [WARNING Δ09s219050us GLib-GObject] invalid class cast from (NULL) pointer to 'GstBin'
Runner: [WARNING Δ06s441502us DioriteGlib] DuplexChannel.vala:342: Channel(2) IOError while receiving data: Failed to read message header. Error receiving data: Connection reset by peer
Runner: [DEBUG Δ000119us DioriteGlib] DuplexChannel.vala:551: Channel (2) has been closed.
Runner: [WARNING Δ000204us Nuvola] AppRunnerController.vala:774: Communication failed: The channel has just been closed.
```
This is probably due to webkit segfault:
```
сер 13 00:20:36 frankenstein.home kernel: WebKitWebProces[7219]: segfault at 210 ip 00007f572efb821e sp 00007ffe1bb84180 error 4 in libwebkit2gtk 4.0.so.37.23.1[7f572d8f7000+2692000]
...
сер 13 00:20:37 frankenstein.home systemd-coredump[7480]: Process 7219 (WebKitWebProces) of user 1000 dumped core.
                                                             
                                                             Stack trace of thread 23:
                                                             #0  0x00007f572efb821e n/a (/app/lib/libwebkit2gtk-4.0.so.37.23.1)
```
Also not sure if this is related but it happens sometimes right before the webkit crash too:
```
сер 13 00:20:15 frankenstein.home systemd-coredump[7115]: Process 7107 (sadc) of user 0 dumped core.
                                                             
                                                             Stack trace of thread 7107:
                                                             #0  0x00007f0ed869d27f __rawmemchr (libc.so.6)
                                                             #1  0x00007f0ed8682292 _IO_str_init_static_internal (libc.so.6)
                                                             #2  0x00007f0ed866ecbd __isoc99_vsscanf (libc.so.6)
                                                             #3  0x00007f0ed866ec24 __isoc99_sscanf (libc.so.6)
                                                             #4  0x000000000040cc42 get_filesystem_nr (sadc)
                                                             #5  0x00000000004044f9 wrap_get_filesystem_nr (sadc)
                                                             #6  0x0000000000402c37 sa_sys_init (sadc)
                                                             #7  0x00000000004021d4 main (sadc)
                                                             #8  0x00007f0ed861f4da __libc_start_main (libc.so.6)
                                                             #9  0x000000000040254a _start (sadc)
```

However, if I press the refresh button in the titlebar and try playing again it sometimes works (sometimes not) but only to  crash again for the next song in the queue. When it works no Flash process seems to be active.

**Comment by excieve** [comment]
Never mind the `sadc` crash — it's unrelated.

**Comment by fenryxo** [comment]
Thanks @excieve. I've rebuilt the latest development WebKitGtk and the lastest GStreamer. Could you check whether the crash still occurs?

**Comment by excieve** [comment]
Updated but it still crashes for me unfortunately:
```
сер 14 12:54:34 frankenstein.home systemd-coredump[21839]: Process 21582 (WebKitWebProces) of user 1000 dumped core.
                                                              
                                                              Stack trace of thread 28:
                                                              #0  0x00007f863aff48be n/a (/app/lib/libwebkit2gtk-4.0.so.37.24.0)
```

As another bit of info it seems I can pretty reliably play songs from my uploaded library (which are mostly mp3s) but crashes mostly happen on random songs from the subscription. Maybe there's a different format that WebKitGtk chokes on or possibly Google Play utilises DRM encryption with EME that it might not support?

**Comment by excieve** [comment]
Just launched Google Play in Chrome and opened the network tab in developer tools. Seems that it doesn't actually use DRM but subscription songs I tried are actually in `audio/mp4` rather than `audio/mpeg`, which might be the thing causing troubles.

**Comment by excieve** [comment]
I think it might actually be streaming in AAC rather than MP3.

**Comment by fenryxo** [comment]
Could you please attach a full log from `flatpak run eu.tiliado.NuvolaAppGooglePlayMusic//master -D`? There might be something helpful.

**Comment by excieve** [comment]
Here it is:
[nuvola_debug.txt](https://github.com/tiliado/nuvolaruntime/files/1223120/nuvola_debug.txt)

**Comment by fenryxo** [comment]
It asks about `audio/mp4; codecs="mp4a.40.2" `:

```
CONSOLE LOG Audio.canPlayType: audio/mp4; codecs="mp4a.40.2" => probably;
  audio/mp4; codecs="mp4a.40.2" => probably
CONSOLE LOG audio/mp4; codecs="mp4a.40.2":
  MediaSource.isTypeSupported => true; Audio.canPlayType => probably;
  result => true
```

**Comment by excieve** [comment]
Well, that's the only difference I can spot to separate playing and crashing songs. Please let me know if there's any way I can help debugging this.

**Comment by excieve** [comment]
I've downloaded a first 4s chunk of the stream for one of the songs. I can test it with something minimal to check if that has to do with one of the components (such as WebKitGtk). Here's the related codec info of this chunk:
```
Stream #0:0(und): Audio: aac (LC) (mp4a / 0x6134706D), 44100 Hz, stereo, fltp, 0 kb/s (default)
```

**Comment by fenryxo** [comment]
I'll try to port Nuvola to use [CEF](https://bitbucket.org/chromiumembedded/cef) instead of WebKitGTK. There should be no issues with MSE.

**Comment by fenryxo** [comment]
Any volunteer to test the work-in-progress CEF backend with Google Play Music? Se also a draft of a blog post [Status of Chromium-based Port of Nuvola](https://medium.com/nuvola-news/status-of-chromium-based-port-of-nuvola-3b1648b29c77).

**Comment by excieve** [comment]
@fenryxo Great stuff! I'll be happy to test it soon.

**Comment by KruyKaze** [comment]
I'll be happy to help. Where do I get it?

**Comment by fenryxo** [comment]
1. Install the **master branch** of [Nuvola Core Developer Kit](https://github.com/tiliado/nuvolaruntime/wiki/Nuvola-Core-Developer-Kit#instalation), but skip any subsequent set-up instructions.
2. Then run `flatpak run --command=Cefium eu.tiliado.NuvolaCdk --disable-flash --url="https://play.google.com/music/"`

**Comment by excieve** [comment]
Works great for me! Although UI feels a little sluggish (no GPU acceleration?)
![screenshot from 2017-12-12 14-03-54](https://user-images.githubusercontent.com/1077193/33883742-a659b41a-df45-11e7-821e-bc2163667268.png)

**Comment by fenryxo** [comment]
> Although UI feels a little sluggish (no GPU acceleration?)

Reticketed as tiliado/valacef#1

**Comment by KruyKaze** [comment]
Runs here with no slowdown but I can't use multimedia hotkeys or nuvolactl
Thanks

**Comment by fenryxo** [comment]
> Runs here with no slowdown but I can't use multimedia hotkeys or nuvolactl

That's because it is not Nuvola but Cefium - a demo web browser to test ValaCEF functionality. Once ValaCEF is complete, it will be used by Nuvola instead of WebKitGTK.

**Comment by fenryxo** [comment]
Blog post: [Status of Chromium-based Port of Nuvola](https://medium.com/nuvola-news/nuvola-chromium-port-status-3b1648b29c77)

**Comment by fenryxo** [comment]
>  _Users with Nuvola Basic/Premium/Patron account will be invited to test development builds soon._
> -- [Status of Chromium-based Port of Nuvola](https://medium.com/nuvola-news/nuvola-chromium-port-status-3b1648b29c77)

All Nuvola Patron & Premium user have been invited to test Nuvola with working Chromium backend, Nuvola Basic users will follow in a week.

**Comment by KruyKaze** [comment]
Looking forward to it!

**Comment by fenryxo** [comment]
>>  _Users with Nuvola Basic/Premium/Patron account will be invited to test development builds soon._
>> -- [Status of Chromium-based Port of Nuvola](https://medium.com/nuvola-news/nuvola-chromium-port-status-3b1648b29c77)
>
> All Nuvola Patron & Premium user have been invited to test Nuvola with working Chromium backend, Nuvola Basic users will follow in a week.

Nuvola Basic users have also been invited.
